api/insight/socket.io.go

```diff
@@ -145,68 +158,104 @@ func (soc *SocketServer) Store(blockData *blockdata.BlockData, msgBlock *wire.Ms
 	// Since the coinbase transaction is generated by the miner, it will never
 	// hit mempool. It must be processed now, with the new block.
 	coinbaseTx := msgBlock.Transactions[0]
-	coinbaseHash := coinbaseTx.TxHash().String()
-	// Check each output's pkScript for subscribed addresses.
-	for _, out := range coinbaseTx.TxOut {
-		_, scriptAddrs, _, err := txscript.ExtractPkScriptAddrs(
-			out.Version, out.PkScript, soc.params)
-		if err != nil {
-			apiLog.Warnf("failed to decode pkScript: %v", err)
-			continue
-		}
-
-		if len(scriptAddrs) == 0 {
-			continue
-		}
-
-		soc.watchedAddresses.RLock()
-		for _, address := range scriptAddrs {
-			addr := address.String()
-			if _, ok := soc.watchedAddresses.c[addr]; ok {
-				apiLog.Debugf(`Broadcasting "%s" in "%s" to subscribed clients.`,
-					addr, coinbaseHash)
-				soc.BroadcastTo(addr, addr, coinbaseHash)
-			}
-		}
-		soc.watchedAddresses.RUnlock()
+	coinbaseHex, err := txhelpers.MsgTxToHex(coinbaseTx)
+	if err != nil {
+		apiLog.Errorf("Failed to decode tx %v: %v", coinbaseTx, err)
+		return err
+	}
 
+	soc.newTxChan <- &NewTx{
+		Hex: coinbaseHex,
 	}
 	return nil
 }
 
-func (soc *SocketServer) sendNewTx(newTxChan chan *NewTx) {
+func (soc *SocketServer) relayNewTxs() {
 	for {
-		ntx, ok := <-newTxChan
+		ntx, ok := <-soc.newTxChan
 		if !ok {
 			break
 		}
 		msgTx, err := txhelpers.MsgTxFromHex(ntx.Hex)
 		if err != nil {
 			continue
 		}
+
+		var vins []InsightSocketVin
+		for _, v := range msgTx.TxIn {
+			txid := v.PreviousOutPoint.Hash.String()
+			idx := v.PreviousOutPoint.Index
+			tree := v.PreviousOutPoint.Tree
+			var addrs []string
+			if txhelpers.IsZeroHashStr(txid) {
+				// Coinbase and stake base inputs need to be "{}".
+				vins = append(vins, InsightSocketVin{})
+				continue
+			} else {
+				addrs, err = txhelpers.OutPointAddressesFromString(
+					txid, idx, tree, soc.txGetter, soc.params)
+				if err != nil {
+					apiLog.Warnf("failed to get outpoint address from txid: %v", err)
+					// Still must append this vin to maintain valid implicit
+					// indexing of vins array.
+				}
+			}
+			vins = append(vins, InsightSocketVin{
+				TxID:      txid,
+				Vout:      idx,
+				Addresses: addrs,
+			})
+		}
+
+		var voutAddrs [][]string
+		for i, v := range msgTx.TxOut {
+			// Allow Vouts to be nil or empty, extracting the addresses from the
+			// pkScripts here.
+			if len(ntx.Vouts) == 0 {
+				_, scriptAddrs, _, err := txscript.ExtractPkScriptAddrs(
+					v.Version, v.PkScript, soc.params)
+				if err != nil {
+					apiLog.Warnf("failed to decode pkScript: %v", err)
+					// still need to append an empty array to voutAddrs.
+				}
+				var addrs []string
+				for i := range scriptAddrs {
+					addrs = append(addrs, scriptAddrs[i].EncodeAddress())
+				}
+				voutAddrs = append(voutAddrs, addrs)
+			} else {
+				voutAddrs = append(voutAddrs, ntx.Vouts[i].ScriptPubKey.Addresses)
```

Comment:
Normal behavior for non-coinbase transactions is for the output addresses to already be decoded in the struct received on `soc.newTxChan`, but for coinbase transactions send from `Store` the `Vouts` slice is empty so we decide the pkScript for each output.